draft-ietf-protected-audience-key-value-services.md

```diff
@@ -0,0 +1,717 @@
+    ? metadata: partitionMetadata,
+    arguments: [* requestArgument],
+    ; One group of keys and common attributes about them
+}
+;Single partition object. A collection of keys that can be processed together.
+
+partitionMetadata = {
+    ? experimentGroupId: tstr,
+    ? slotSize: tstr,
+    ? allSlotsRequestedSizes: tstr,
+}
+
+requestArgument = {
+    ? tags: [* tstr],
+    ; List of tags describing this group's attributes. These MAY be picked from the list of available tags in {{tags}}.
+    ? data: [* tstr],
+    ; List of keys to get values for
+}
+~~~~~
+
+#### Available Tags {#tags}
+
+Each key group has exactly one tag from the `namespace` category.
+
+| Tag | Description |
+|---|---|
+| interestGroupNames | Names of interest groups in the encompassing partition. |
+| keys | "keys" represent the keys to be looked up from the service's internal datastore. |
+| renderUrls | "renderUrls" represent URLs for advertisements to be looked up from the service's internal datastore. |
+| adComponentRenderUrls | "adComponentRenderUrls" represent component URLs for advertisements to be looked up from the service's internal datastore. |
+
+
+### Generating a Request
+
+TODO
+
+### Parsing a Request {#request-parsing}
+
+This section describes how the Key Value Service MUST deserialize request messages from the client.
+
+The algorithm takes as input a serialized request message from the client and a list of HPKE private
+keys (along with their corresponding key IDs).
+
+The output is either an error sent back to the client, an empty message sent back to the client, or
+a request message the Key Value Service can consume along with an HPKE context.
+
+1. Let `encrypted request` be the request received from the client.
+1. Let `error_msg` be an empty string.
+1. Decrypt `encrypted request` by using the input private key corresponding to
+   `key_id` as described in {{request-encryption}}, to get the decrypted message and `rctxt`.
+    1. If decryption fails, return failure.
+    2. Else, save the decrypted output as `framed request` and save `rctxt`.
+2. Remove and extract the first 5 bytes from `framed request` as the `framing header` (described in
+   {{framing}}), removing them from `framed request`.
+    1. If the `framing header`'s `Compression` field is not `0x00` (no compression), return failure.
+3. Let `length` be equal to the `framing header`'s `Size` field.
+4. If `length` is greater than the length of the remaining bytes in `framed request`, return
+   failure.
+5. Take the first `length` remaining bytes in `framed response` as `decodable request`, discarding
+   the rest.
+6. [CBOR] decode `decodable request` into the message represented in {{request-schema}}. Let this be
+   `processed request`.
+    1. If decoding fails, return failure.
+7. If no `partitions` are present, return failure.
```

Comment:
Should the schema specify that at least one "partitions" is required (e.g. `partitions: [1* partition]`)?